crates/cli-support/src/interpret.rs

```diff
@@ -0,0 +1,155 @@
+//! basically as a mapping of rustc's trait resolution in executable code. This
+//! allows us to detect, after the macro is invoke, what trait selection did and
+//! what types of functions look like. By executing descriptor functions they'll
+//! each invoke a known import (with only one argument) some number of times,
+//! which gives us a list of `u32` values to then decode.
+//!
+//! The interpreter here is only geared towards this one exact use case, so it's
+//! quite small and likely not extra-efficient.
+
+use std::collections::HashMap;
+use parity_wasm::elements::*;
+
+#[derive(Default)]
+pub struct Interpreter {
+    imports: usize,
+    describe_idx: Option<u32>,
+    name_map: HashMap<String, u32>,
+    code_idx: Option<usize>,
+    sp: i32,
+    mem: Vec<i32>,
+    stack: Vec<i32>,
+    descriptor: Vec<u32>,
+}
+
+impl Interpreter {
+    pub fn new(module: &Module) -> Interpreter {
+        let mut ret = Interpreter::default();
+        ret.mem = vec![0; 0x100];
+        ret.sp = ret.mem.len() as i32;
+
+        for (i, s) in module.sections().iter().enumerate() {
+            match s {
+                Section::Code(_) => ret.code_idx = Some(i),
+                _ => {}
+            }
+        }
+
+        if let Some(i) = module.import_section() {
+            ret.imports = i.functions();
+            for (i, entry) in i.entries().iter().enumerate() {
+                if entry.module() != "__wbindgen_placeholder__" {
+                    continue
+                }
+                if entry.field() != "__wbindgen_describe" {
+                    continue
+                }
+                ret.describe_idx = Some(i as u32);
+            }
+        }
+
+        if let Some(e) = module.export_section() {
+            for e in e.entries() {
+                let i = match e.internal() {
+                    Internal::Function(i) => i,
+                    _ => continue,
+                };
+                ret.name_map.insert(e.field().to_string(), *i);
+            }
+        }
+
+        return ret
+    }
+
+    pub fn interpret(&mut self, func: &str, module: &Module) -> Option<&[u32]> {
+        self.descriptor.truncate(0);
+        let idx = match self.name_map.get(func) {
+            Some(i) => *i,
+            None => panic!("{} not found in exported module", func),
+        };
+        let code = match &module.sections()[self.code_idx.unwrap()] {
+            Section::Code(s) => s,
+            _ => panic!(),
+        };
+        assert_eq!(self.sp, self.mem.len() as i32);
+        assert_eq!(self.stack.len(), 0);
+        self.call(idx, code);
+        assert_eq!(self.stack.len(), 0);
+        assert_eq!(self.sp, self.mem.len() as i32);
+        Some(&self.descriptor)
+    }
+
+    fn call(&mut self, idx: u32, code: &CodeSection) {
+        use parity_wasm::elements::Instruction::*;
+
+        let body = &code.bodies()[(idx as usize) - self.imports];
```

Comment:
Assert that we aren't invoking an imported type (and it shouldn't be the describe function either because of the inline handling of it below).

```rust
assert!(idx >= self.imports);
```